src/main.rs

```diff
@@ -47,125 +87,249 @@ impl LogLevel {
-        for entry in glob(path)?.flatten() {
+    Ok(())
+}
+
+/// runs the scripts green.d when health-check passes
+fn run_green() -> Result<(), Error> {
+    for path in GREENBOOT_INSTALL_PATHS {
+        let green_path = format!("{path}/green.d/*.*");
+        for entry in glob(&green_path)?.flatten() {
             log::info!("running green check {}", entry.to_string_lossy());
-            let output = Command::new("bash").arg("-C").arg(entry).output()?;
+            let output = Command::new("bash").arg("-C").arg(entry.clone()).output()?;
             if !output.status.success() {
                 // combine and print stderr/stdout
-                log::warn!("green script failed...");
+                log::warn!("green script {} failed!", entry.to_string_lossy());
+                log::warn!("reason: {}", String::from_utf8_lossy(&output.stderr));
             }
         }
     }
     Ok(())
 }
 
-fn reboot() -> Result<(), Error> {
-    Command::new("systemctl").arg("reboot").spawn()?;
-    Ok(())
+/// triggers the diagnostics followed by the action on the outcome
+/// this also handle setting the grub variables and system restart
+fn health_check() -> Result<()> {
+    let config = GreenbootConfig::get_config();
+    log::info!("{config:?}");
+    handle_motd("healthcheck is in progress")?;
+    let run_status = run_diagnostics();
+    match run_status {
+        Ok(()) => {
+            log::info!("greenboot health-check passed.");
+            run_green().unwrap_or_else(|e| {
+                log::error!("cannot run green script due to: {}", e.to_string())
+            });
+            handle_motd("healthcheck passed - status is GREEN")
+                .unwrap_or_else(|e| log::error!("cannot set motd due to : {}", e.to_string()));
+            handle_boot_success(true)?;
+            Ok(())
+        }
+        Err(e) => {
+            log::error!("Greenboot health-check failed!");
+            handle_motd("healthcheck failed - status is RED")
+                .unwrap_or_else(|e| log::error!("cannot set motd due to : {}", e.to_string()));
+            run_red()
+                .unwrap_or_else(|e| log::error!("cannot run red script due to: {}", e.to_string()));
+            handle_boot_success(false)?;
+            set_boot_counter(config.max_reboot)
+                .unwrap_or_else(|e| log::error!("cannot set boot_counter as: {}", e.to_string()));
+            handle_reboot(false)
+                .unwrap_or_else(|e| log::error!("cannot reboot as: {}", e.to_string()));
+            Err(e)
+        }
+    }
 }
 
-fn stamp() -> Result<(), Error> {
-    fs::create_dir_all("/etc/greenboot/")?;
-    File::create("/etc/greenboot/upgrade.stamp")?;
-    Ok(())
+/// initiates rollback if boot_counter and satisfies
+fn trigger_rollback() -> Result<()> {
+    match handle_rollback() {
+        Ok(()) => {
+            log::info!("Rollback successful");
+            unset_boot_counter()?;
+            handle_reboot(true)?;
+            Ok(())
```

Comment:
```suggestion
            handle_reboot(true)
```
we shouldn't mask this